### DISCUSSION TITLE
Feature request - Snippets

### DISCUSSION THREAD

**Discussion by willblackmore** [outline/outline#2030] [Ideas]
We are currently populating internal documentation for our business, and one challenge we are facing is reducing duplication of content across mostly similar documents.

For example, we have a comprehensive collection of job specifications for every role in our business. While each one is different overall many share common content (eg: senior responsibilities are the same for all senior roles) which we are currently copy+pasting.

Obviously, this gets hard to maintain, really quickly.

So, I'd like to suggest a new 'snippets' feature. The idea being that you create and manage snippets of content (with all the usual formatting options) within a snippets area. These can then be included in another document using the usual UI. When viewing a document, you'd never know it was a snippet, but while editing, it could be faded and show a message stating 'Edit this snippet'.

Thoughts?

**Comment by tommoor** [comment]
Ah yes, there was a similar request here: https://github.com/outline/outline/issues/708 – are you using share links from Outline to share your job listings in this case?

Is the idea that editing the snippet would update in any document that includes it – e.g. it's dynamic? While I see the ways this could be useful I do have a couple of concerns with such a feature:

- I think this will be quite difficult to communicate to the end user what's going, there would be an island of content that you can't edit and how this appears would need careful consideration.
- The performance impact of needing to load N snippets to display a document would probably be noticable, particularly if folks put multiple different snippets in a document
- We're moving towards realtime editing and that would complicate the implementation of this also.

**Reply by willblackmore** [reply]
> are you using share links from Outline to share your job listings in this case?

No, not in this case. They are for internal reference.

> Is the idea that editing the snippet would update in any document that includes it

That's what I'm thinking yes!

> I think this will be quite difficult to communicate to the end user what's going, there would be an island of content that you can't edit and how this appears would need careful consideration.

I thought this too. I was wondering if in editing mode the snippet block could be faded out with an overlay stating that it is a snippet and an 'Edit this snippet' CTA.

<img width="941" alt="Editor" src="https://user-images.githubusercontent.com/1260445/114839345-92b31b00-9dcd-11eb-9dc9-11ab3cedbceb.png">

> The performance impact of needing to load N snippets to display a document would probably be noticeable, particularly if folks put multiple different snippets in a document

Instead of loading the snippet from its origin each time, could the system duplicate the contents of the snippet into the page when it is added and then create a relationship between the snippet and page(s) it is included on?

When viewing a file, I'd imagine there would be no performance impact at all as it is still loading content from one location as it does now. However, when someone edits a snippet the system would need to loop through the various relationships and update instances of itself in each page. I imagine that would impact performance, but it is only when someone performs a specific action (which could come with a warning, a bit like the export feature), rather than inadvertently just by browsing.

**Some other examples for the UI**

<img width="824" alt="Landing" src="https://user-images.githubusercontent.com/1260445/114839352-93e44800-9dcd-11eb-859a-e6a5e9d64ed9.png">

<img width="313" alt="Nav" src="https://user-images.githubusercontent.com/1260445/114839357-947cde80-9dcd-11eb-8259-f18ef5ca3f5d.png">

**Reply by tommoor** [reply]
Thanks for the thoughtful response and mocks, it definitely helps in the feature consideration

**Comment by kyleshrader** [comment]
I think a feature named snippets should be similar to templates, but for insertable templatized content. It should also include placeholder support. Much like a vscode or intellij "snippet". I will be suggesting this feature, but for this snippets idea I would suggest a name like "excerpts" or "references" (similar to the mentioned #708).

**Reply by tommoor** [reply]
I want to enable templates to be inserted anywhere in a document through the /slash menu, I feel like this would probably satisfy the snippet usecase without the creation of a new entity?

**Comment by andrew-chen-2202** [comment]
This feature is also very desirable for our team to manage external-facing, templatized documents (Job Description). Is there any plan to put it on the roadmap @tommoor. Again, thanks for building this exceptional application.

**Comment by EduNetEurope** [comment]
Coming from Archbee, this is also a feature we desperately miss. Archbee's "Content Snipps" are brilliant.

**Comment by EduNetEurope** [comment]
Same request here: https://github.com/outline/outline/discussions/7164

**Comment by AndrewBucklin** [comment]
This would be very helpful for our team. We manage multiple clients and we tend to reuse certain steps / processes for each client. Sometimes these steps charge or need to be updated and we constantly find ourselves following various revisions of those steps since we tend to forget to update the steps manually in all located throughout the Wiki. One workaround we have tried to implement is storing those common steps in a Google Doc and linking to that (or embedding it), but it's clunky and goes against the idea of having all documentation within a single system.
